packages/blocks/src/store/actions.js

```diff
@@ -1,191 +1,253 @@
 /**
  * External dependencies
  */
-import { castArray } from 'lodash';
+import {
+	castArray,
+	keyBy,
+	map,
+	omit,
+	mapValues,
+	uniqBy,
+	find,
+	isEmpty,
+} from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	blockTypesByName,
+	blockStylesByBlockName,
+	blockVariationsByBlockName,
+	defaultBlockName,
+	freeformFallbackBlockName,
+	unregisteredFallbackBlockName,
+	groupingBlockName,
+	blockCategories,
+	blockCollections,
+} from './atoms';
 
 /** @typedef {import('../api/registration').WPBlockVariation} WPBlockVariation */
 
 /**
  * Returns an action object used in signalling that block types have been added.
  *
- * @param {Array|Object} blockTypes Block types received.
- *
- * @return {Object} Action object.
+ * @param {Array|Object} newBlockTypes Block types received.
  */
-export function addBlockTypes( blockTypes ) {
-	return {
-		type: 'ADD_BLOCK_TYPES',
-		blockTypes: castArray( blockTypes ),
-	};
-}
+export const addBlockTypes = ( newBlockTypes ) => ( { get, set } ) => {
+	newBlockTypes = castArray( newBlockTypes );
+	const currentBlockTypes = get( blockTypesByName );
+	set( blockTypesByName, {
+		...currentBlockTypes,
+		...keyBy(
+			map( newBlockTypes, ( blockType ) => omit( blockType, 'styles ' ) ),
+			'name'
+		),
+	} );
+
+	const currentBlockStyles = get( blockStylesByBlockName );
+	set( blockStylesByBlockName, {
+		...currentBlockStyles,
+		...mapValues( keyBy( newBlockTypes, 'name' ), ( blockType ) => {
+			return uniqBy(
+				[
+					...( blockType.styles ?? [] ),
+					...( currentBlockStyles[ blockType.name ] ?? [] ),
+				],
+				( style ) => style.name
+			);
+		} ),
+	} );
+
+	const currentBlockVariations = get( blockVariationsByBlockName );
+	set( blockVariationsByBlockName, {
+		...currentBlockVariations,
+		...mapValues( keyBy( newBlockTypes, 'name' ), ( blockType ) => {
+			return uniqBy(
+				[
+					...( blockType.variations ?? [] ),
+					...( currentBlockVariations[ blockType.name ] ?? [] ),
+				],
+				( variation ) => variation.name
+			);
+		} ),
+	} );
+};
```

Comment:
if you look closely, you'll notice that basically, the logic that was in the reducers, spread across multiple reducers is now inside the action that triggered the change. I like the explicitness personally, that said, if we want to share the same logic across actions, we should be able to move this logic to the "set" function of the atoms.

Reply:
In this particular store, I didn't feel the need to share behavior between actions.